ambari-server/src/main/java/org/apache/ambari/server/security/encryption/ConfigPropertiesEncryptor.java

```diff
@@ -0,0 +1,127 @@
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * {@link Encryptor} implementation for encrypting/decrypting PASSWORD type
+ * properties in {@link Config}'s properties
+ */
+
+@Singleton
+public class ConfigPropertiesEncryptor implements Encryptor<Config> {
+
+  private static final String ENCRYPTED_PROPERTY_PREFIX = "${enc=aes256_hex, value=";
+  private static final String ENCRYPTED_PROPERTY_SCHEME = ENCRYPTED_PROPERTY_PREFIX + "%s}";
+
+  private final EncryptionService encryptionService;
+  private final Map<Long, Map<StackId, Map<String, Set<String>>>> clusterPasswordProperties = new ConcurrentHashMap<>(); //Map<clusterId, <Map<stackId, Map<configType, Set<passwordPropertyKeys>>>>;
+
+  @Inject
+  public ConfigPropertiesEncryptor(EncryptionService encryptionService) {
+    this.encryptionService = encryptionService;
+  }
+
+  @Override
+  public void encryptSensitiveData(Config config) {
+    try {
+      final Map<String, String> configProperties = config.getProperties();
+      if (configProperties != null) {
+        final Set<String> passwordProperties = getPasswordProperties(config.getCluster(), config.getType());
+        if (CollectionUtils.isNotEmpty(passwordProperties)) {
+          final Map<String, String> encryptedProperties = new HashMap<>(configProperties);
+          for (Map.Entry<String, String> property : configProperties.entrySet()) {
+            if (passwordProperties.contains(property.getKey()) && !isEncryptedPassword(property.getValue())) {
+              encryptedProperties.put(property.getKey(), encryptAndDecoratePropertyValue(property.getValue()));
+            }
+          }
+          config.setProperties(encryptedProperties);
+        }
+      }
+    } catch (Exception e) {
+      throw new AmbariRuntimeException("Error while encrypting sensitive data", e);
+    }
+  }
+
+  private boolean isEncryptedPassword(String password) {
+    return password != null && password.startsWith(ENCRYPTED_PROPERTY_PREFIX); // assuming previous encryption by this class
+  }
+
+  private Set<String> getPasswordProperties(Cluster cluster, String configType) throws AmbariException {
+    //in case of normal configuration change on the UI - or via the API - the current and desired stacks are equal
+    //in case of an upgrade they are different; in this case we want to get password properties from the desired stack
+    if (cluster.getCurrentStackVersion().equals(cluster.getDesiredStackVersion())) {
+      return getPasswordProperties(cluster, cluster.getCurrentStackVersion(), configType);
+    } else {
+      return getPasswordProperties(cluster, cluster.getDesiredStackVersion(), configType);
+    }
+  }
+
+  private Set<String> getPasswordProperties(Cluster cluster, StackId stackId, String configType) {
+    final long clusterId = cluster.getClusterId();
+    clusterPasswordProperties.computeIfAbsent(clusterId, v -> new ConcurrentHashMap<>()).computeIfAbsent(stackId, v -> new ConcurrentHashMap<>())
```

Comment:
Why bother caching using the cluster id at all.  Even if Ambari was managing multiple clusters, the stack definitions all would be the same for the same stack id.

Reply:
hm..that's true. I may remove that layer next week.